python/rmm/_lib/device_buffer.pxd

```diff
@@ -38,6 +39,15 @@ cdef extern from "rmm/device_buffer.hpp" namespace "rmm" nogil:
 cdef class DeviceBuffer:
     cdef unique_ptr[device_buffer] c_obj
 
+    # Holds a reference to the MemoryResource used for allocation. Ensures the
+    # MR does not get destroyed before this DeviceBuffer. `mr` is needed for
+    # deallocation
+    cdef MemoryResource mr
+
+    # Holds a reference to the stream used by the underlying `device_buffer`.
+    # Ensures the stream does not get destroyed before this DeviceBuffer
+    cdef Stream stream
```

Comment:
Shouldn't these also be held in `unique_ptr`s?

Reply:
No -- the `DeviceBuffer` is not the sole owner of the `MemoryResource` or the `Stream`. It just holds references to them so that they can't get destroyed out from under it.

Reply:
Do those objects owning `unique_ptr`s then? Just trying to understand how they get cleaned up.

Reply:
Yes, they could have unique_ptrs underneath them

Reply:
I say "could" because a Stream could have a `unique_ptr[RMM::cuda_stream` underneath it, but it could also have a numba or CuPy stream instead. Each of those are "owning" objects.

Reply:
These are Cython classes as opposed to C++ classes, so they'll still go through Python refcounting

Reply:
As Keith said, these go through the Python ref counting system. Resolving